torch/_inductor/optimize_indexing.py

```diff
@@ -63,247 +36,85 @@ def range_expressable_in_32_bits(range):
-    kernels such as Upsample and Interpolate.
-    """
-
-    def __init__(
-        self,
-        loop_body: LoopBody,
-        indices_ranges: Dict[sympy.Symbol, int],
-        indexing_exprs: Dict[str, sympy.Expr],
-    ):
-        self.loop_body = loop_body
-        self.indices_range = indices_ranges
-        self.indexing_exprs = indexing_exprs
-        self.replacement_vals = {}
-        self.interp_env = {}
-        self.submodules = self.swap_submodules(dict(loop_body.submodules))
-
-        indirect_var_set = set(loop_body.indirect_vars)
-        self.index_indirect_dependecies = {
-            index: expr.free_symbols & indirect_var_set
-            for index, expr in indexing_exprs.items()
-        }
-        self.all_graphs = [loop_body.root_block.graph] + [
-            block.graph for block in loop_body.subblocks.values()
-        ]
-
-        for k, v in indices_ranges.items():
-            self.replace_indirect(k, ValueRanges(0, v))
-
-        # avoid computing these values, pessimistically assume that they are unbounded
-        self.tensor_values_set = dominated_nodes(
-            [
-                node
-                for node in self.all_nodes
-                if node.target in ["load", "reduction"]
-                or "masked_subblock" in node.target
-            ]
+def try_to_reduce_precision(node, bounds, indirect_vars, indices, replacement_vals):
+    # if a downstream use of a node explicitly converts to int32, or float16/float32/float64,
+    # then it's precision is set for that chain of uses, and we don't need to consider those
+    # dominated values
+    def skip_filter(node):
+        return node.target == "to_dtype" and node.args[2] in (
+            torch.int32,
+            torch.float32,
+            torch.float64,
         )
 
-    def run(self):
-        """Compute Value Ranges and try reduce precision of 'to_dtype' nodes to int32 where possible"""
-
-        int64_dtype_nodes = [
-            node
-            for node in self.all_nodes
-            if (
-                node.target == "to_dtype"
-                and node.args[2] == torch.int64
-                and node not in self.tensor_values_set
-            )
-        ]
-        if not int64_dtype_nodes:
+    # TODO - there are dominated uses whose dtype does not depend on whether
+    # we reduce the precision here, e.g. add(int64, int64) one of the args can be reduced to
+    # int32 without changing the output precision of the node. this case hasn't shown up
+    for dominated in dominated_nodes([node], skip_filter):
+        if dominated.target in ["store", "output"]:
+            continue
+
+        if "set_indirect" in dominated.target:
+            idx = int(dominated.target[len("set_indirect") :])
+            indirect_var = indirect_vars[idx]
+
+            # We check that we can compute all the indices it's involved in with int32
+            # TODO The current implementation does not handle the case were an indexing
```

Comment:
where did the new TODO come up ?

Reply:
It just occurred to me when having a look at this code. This situation comes up quite often when you have different tensors indexing different dimensions. For example, in the triton code generated in the OP of https://github.com/pytorch/pytorch/issues/100530